src/components/ha-data-table.ts

```diff
@@ -241,29 +296,94 @@ export class HaDataTable extends BaseElement {
 
   private _handleRowCheckboxChange(ev: Event) {
     const checkbox = ev.target as HaCheckbox;
-    const rowIndex = Number(
-      checkbox.parentElement!.parentElement!.getAttribute("data-row-id")
+    const rowId = checkbox.parentElement!.parentElement!.getAttribute(
+      "data-row-id"
     );
 
-    this._setRowCheckboxCheckedAtIndex(rowIndex!, checkbox.checked);
+    this._setRowChecked(rowId!, checkbox.checked);
     this.mdcFoundation.handleRowCheckboxChange(ev);
   }
 
-  private _setRowCheckboxCheckedAtIndex(rowIndex: number, checked: boolean) {
-    if (checked && !this._checkedRows.includes(rowIndex)) {
-      this._checkedRows = [...this._checkedRows, rowIndex];
+  private _setRowChecked(rowId: string, checked: boolean) {
+    if (checked && !this._checkedRows.includes(rowId)) {
+      this._checkedRows = [...this._checkedRows, rowId];
     } else if (!checked) {
-      const index = this._checkedRows.indexOf(rowIndex);
+      const index = this._checkedRows.indexOf(rowId);
       if (index !== -1) {
         this._checkedRows.splice(index, 1);
       }
     }
     fireEvent(this, "selection-changed", {
-      index: rowIndex,
+      id: rowId,
       selected: checked,
     });
   }
 
+  private _handleSearchChange(ev: CustomEvent): void {
+    this._filter = ev.detail.value;
+  }
+
+  private _filterData(data: DataTabelRowData[], filter: string) {
+    if (!filter) {
+      return data;
+    }
+    const ucFilter = filter.toUpperCase();
+    return data.filter((row) => {
+      return Object.entries(this.columns).some((columnEntry) => {
```

Comment:
If something is used inside memoized, all dependencies should be passed in as parameters, otherwise the memoization is not working correctly. I would suggest you define this method inside the `memoizeOne` function to make it clear that it is memoized.